pygfx/utils/transform.py

```diff
@@ -0,0 +1,449 @@
+import numpy as np
+import pylinalg as la
+from time import perf_counter_ns
+import weakref
+import functools
+
+from typing import List, Tuple, Union
+
+
+class cached:  # noqa: N801
+    """Cache for computed properties.
+
+    This descriptor implements a minimal timestamp-based cache for computed
+    properties. The value of the property is computed using ``compute_fn`` and
+    the result is cached until ``obj.last_modified`` advances. At this point the
+    value of the computed property is recomputed upon the next read.
+
+    """
+
+    def __init__(self, compute_fn=None) -> None:
+        self.compute_fn = compute_fn
+        self.name = None
+
+    def __set_name__(self, clazz, name) -> None:
+        self.name = f"_{name}_cache"
+
+    def __get__(self, instance: "AffineBase", clazz=None) -> np.ndarray:
+        if instance is None:
+            return self
+
+        if not hasattr(instance, self.name):
+            cache = (instance.last_modified, self.compute_fn(instance))
+            setattr(instance, self.name, cache)
+        else:
+            cache = getattr(instance, self.name)
+
+        if instance.last_modified > cache[0]:
+            cache = (instance.last_modified, self.compute_fn(instance))
+            setattr(instance, self.name, cache)
+
+        return cache[1]
+
+
+class callback:  # noqa: N801
+    """Weakref support for AffineTransform callbacks.
+
+    This decorator replaces the instance reference of a class method (self) with
+    a weakref to the instance. It also dynamically resolves this weakref
+    so that users don't have to.
+
+    The use-case for it is to avoid the circular reference that happens when a
+    WorldObject registers a callback to keep it's local transform in sync with
+    its world transform. This speeds up garbage collection as WorldObjects will
+    get removed once the last reference is deleted instead of waiting for GC's
+    cycle detection phase.
```

Comment:
@almarklein is this conceptually equivalent to the weakref callbacks system you implemented in wgpu.gui ? we might be able to reuse this decorator there to clean things up?

Reply:
This one?
https://github.com/pygfx/wgpu-py/blob/402961a04299a1518ac0fb7ad3dbb44636679dbb/wgpu/gui/base.py#L45-L57

It looks like the purpose is indeed similar, except implement in a functional approach instead of a descriptor.

We could reuse that by `from wgpu.gui.base import weakbind`, but that means we import wgpu.gui` even when no gui is being used ... For these 10 lines of code I think its fine to just copy / re-implement.